# Pandas Series

A Pandas Series is like a column in a table. It is a 1-D array holding any type of data

In [1]:
import numpy as np
import pandas as pd

## Series from list

In [2]:
#string
country = ['Nepal','India','China','USA']
pd.Series(country)  #Series is class in pandas

0    Nepal
1    India
2    China
3      USA
dtype: object

In [3]:
#integers
goals = [1,2,3,1,0,0,1,2,3]
pd.Series(goals)

0    1
1    2
2    3
3    1
4    0
5    0
6    1
7    2
8    3
dtype: int64

in series there are two items(one is index and another is value(that we provide))

In [4]:
#custom index
players = ['ronaldo','messi','lewandowski','benzema','neymar']
goals = [140,129,103,90,43]
pd.Series(goals,index = players)

ronaldo        140
messi          129
lewandowski    103
benzema         90
neymar          43
dtype: int64

In [5]:
#setting a name
pd.Series(goals,index = players,name = 'Football players ucl goals')

ronaldo        140
messi          129
lewandowski    103
benzema         90
neymar          43
Name: Football players ucl goals, dtype: int64

## Series from dictinary

In [6]:
goals = {
    'ronaldo':140,
    'messi':129,
    'lewandowski':103,
    'benzema':90,
    'neymar':43
}

In [10]:
goals_series = pd.Series(goals,name="Top 5 UCL Scorer")
goals_series

ronaldo        140
messi          129
lewandowski    103
benzema         90
neymar          43
Name: Top 5 UCL Scorer, dtype: int64

## Series Attributes

In [11]:
#size
goals_series.size

5

In [12]:
#dtype
goals_series.dtype

dtype('int64')

In [13]:
#name
goals_series.name

'Top 5 UCL Scorer'

In [16]:
#is_unique   ----> very useful
goals_series.is_unique

True

In [15]:
#index --------> very useful
goals_series.index

Index(['ronaldo', 'messi', 'lewandowski', 'benzema', 'neymar'], dtype='object')

In [17]:
#values ---------------> very very useful
goals_series.values

array([140, 129, 103,  90,  43], dtype=int64)

## series using read_csv

read_csv is one of the most used and important function to load a csv format data

In [19]:
#with one column
type(pd.read_csv('../data/subs.csv'))

pandas.core.frame.DataFrame

!!! when pd.read_csv loads and stores data it stores in the format of DataFrame not series !!!

we can convert dataframe into series using squeeze() method

In [21]:
data = pd.read_csv('../data/subs.csv')
data = data.squeeze()
print(data)
type(data)

0       48
1       57
2       40
3       43
4       44
      ... 
360    231
361    226
362    155
363    144
364    172
Name: Subscribers gained, Length: 365, dtype: int64


pandas.core.series.Series

In [23]:
# with 2 columns
vk = pd.read_csv('../data/kohli_ipl.csv')
vk

,match_no,runs
0,1,1
1,2,23
2,3,13
3,4,12
4,5,1
...,...,...
210,211,0
211,212,20
212,213,73
213,214,25


it has two columns one is match_no and another is runs

In [24]:
type(vk)

pandas.core.frame.DataFrame

it is still pandas DataFrame, lets conver it into series as :

In [28]:
vk = pd.read_csv('../data/kohli_ipl.csv',index_col='match_no')
vk = vk.squeeze()
vk

match_no
1       1
2      23
3      13
4      12
5       1
       ..
211     0
212    20
213    73
214    25
215     7
Name: runs, Length: 215, dtype: int64

In [29]:
type(vk)

pandas.core.series.Series

In [32]:
#lets load another data
bollywood = pd.read_csv('../data/bollywood.csv',index_col='movie')
bollywood

,lead
movie,
Uri: The Surgical Strike,Vicky Kaushal
Battalion 609,Vicky Ahuja
The Accidental Prime Minister (film),Anupam Kher
Why Cheat India,Emraan Hashmi
Evening Shadows,Mona Ambegaonkar
...,...
Hum Tumhare Hain Sanam,Shah Rukh Khan
Aankhen (2002 film),Amitabh Bachchan
Saathiya (film),Vivek Oberoi


In [33]:
type(bollywood)

pandas.core.frame.DataFrame

In [36]:
bollywood = bollywood.squeeze()

In [37]:
type(bollywood)

pandas.core.series.Series

## Series Methods

In [38]:
#head
data.head()

0    48
1    57
2    40
3    43
4    44
Name: Subscribers gained, dtype: int64

In [39]:
#tail
data.tail()

360    231
361    226
362    155
363    144
364    172
Name: Subscribers gained, dtype: int64

In [41]:
data.head(7)   #first 7 rows,similar for tail

0    48
1    57
2    40
3    43
4    44
5    46
6    33
Name: Subscribers gained, dtype: int64

In [64]:
#sample()             ----> this method returns random one row from series
bollywood.sample()

movie
Dhoom 3    Aamir Khan
Name: lead, dtype: object

In [65]:
#value_counts ---------> very very useful function , returns the frequency of values
bollywood

movie
Uri: The Surgical Strike                   Vicky Kaushal
Battalion 609                                Vicky Ahuja
The Accidental Prime Minister (film)         Anupam Kher
Why Cheat India                            Emraan Hashmi
Evening Shadows                         Mona Ambegaonkar
                                              ...       
Hum Tumhare Hain Sanam                    Shah Rukh Khan
Aankhen (2002 film)                     Amitabh Bachchan
Saathiya (film)                             Vivek Oberoi
Company (film)                                Ajay Devgn
Awara Paagal Deewana                        Akshay Kumar
Name: lead, Length: 1500, dtype: object

In [66]:
bollywood.value_counts()

lead
Akshay Kumar        48
Amitabh Bachchan    45
Ajay Devgn          38
Salman Khan         31
Sanjay Dutt         26
                    ..
Diganth              1
Parveen Kaur         1
Seema Azmi           1
Akanksha Puri        1
Edwin Fernandes      1
Name: count, Length: 566, dtype: int64

we can clearly see in the data bollywood, akshay kumar is in lead role in 48 movies

In [67]:
#sort_values
vk

match_no
1       1
2      23
3      13
4      12
5       1
       ..
211     0
212    20
213    73
214    25
215     7
Name: runs, Length: 215, dtype: int64

In [68]:
vk.sort_values()

match_no
87       0
211      0
207      0
206      0
91       0
      ... 
164    100
120    100
123    108
126    109
128    113
Name: runs, Length: 215, dtype: int64

from this we can see virat kohli scores 113 runs in his 126th ipl match

In [69]:
vk.sort_values(ascending=False)  #descending sort

match_no
128    113
126    109
123    108
164    100
120    100
      ... 
93       0
211      0
130      0
8        0
135      0
Name: runs, Length: 215, dtype: int64

Finding and printing kohli`s highest run as :

In [70]:
vk.sort_values(ascending=False).head(1).values[0]

113

In [71]:
vk.sort_values(ascending=False)

match_no
128    113
126    109
123    108
164    100
120    100
      ... 
93       0
211      0
130      0
8        0
135      0
Name: runs, Length: 215, dtype: int64

In [72]:
vk

match_no
1       1
2      23
3      13
4      12
5       1
       ..
211     0
212    20
213    73
214    25
215     7
Name: runs, Length: 215, dtype: int64

i have sorted vk but this is temporary change it is not reflected in original data(vk), we can do changes permanent using inplace

In [75]:
#sort_index
bollywood.sort_index()

movie
1920 (film)                   Rajniesh Duggall
1920: London                     Sharman Joshi
1920: The Evil Returns             Vicky Ahuja
1971 (2007 film)                Manoj Bajpayee
2 States (2014 film)              Arjun Kapoor
                                   ...        
Zindagi 50-50                      Veena Malik
Zindagi Na Milegi Dobara        Hrithik Roshan
Zindagi Tere Naam           Mithun Chakraborty
Zokkomon                       Darsheel Safary
Zor Lagaa Ke...Haiya!            Meghan Jadhav
Name: lead, Length: 1500, dtype: object

In [78]:
bollywood.sort_index(inplace=True)  #inplace for permanent change to origial
bollywood

movie
1920 (film)                   Rajniesh Duggall
1920: London                     Sharman Joshi
1920: The Evil Returns             Vicky Ahuja
1971 (2007 film)                Manoj Bajpayee
2 States (2014 film)              Arjun Kapoor
                                   ...        
Zindagi 50-50                      Veena Malik
Zindagi Na Milegi Dobara        Hrithik Roshan
Zindagi Tere Naam           Mithun Chakraborty
Zokkomon                       Darsheel Safary
Zor Lagaa Ke...Haiya!            Meghan Jadhav
Name: lead, Length: 1500, dtype: object

## Series maths Methods

In [79]:
#count -------> same as size but does not include missing values
vk.count()

215

In [80]:
#sum --------> total sum of values
vk.sum()

6634

kohli scores 6634 runs in ipl in 215 games

In [82]:
#mean
vk.mean()  

30.855813953488372

average of virat kohli is 30.85 per game

In [83]:
#median
vk.median()

24.0

In [84]:
#standard deviation
vk.std()

26.22980132830278

In [85]:
#variance
vk.var()

688.0024777222344

In [86]:
#mode
bollywood.mode()

0    Akshay Kumar
Name: lead, dtype: object

In [87]:
#min
vk.min()

0

In [88]:
#max
vk.max()

113

In [89]:
#describe -----> important,gives mathemaical summary
vk.describe()

count    215.000000
mean      30.855814
std       26.229801
min        0.000000
25%        9.000000
50%       24.000000
75%       48.000000
max      113.000000
Name: runs, dtype: float64

## Series Indexing